### DISCUSSION TITLE
[Question][Hw3] Several questions

### DISCUSSION THREAD

**Discussion by kws1207** [prosyslab-classroom/cs348-information-security#520] [Q&A]
1. Can you please give us a hint about how to use declare_function? or at least example code? I don't know how to define function_type.

2. How does the declare_function work? Does it in some way automatically finds external c function?

3. What is llctx, and what is llm? What are the type of them?

4. In Ocaml, how can we make a single function to do two things?

**Comment by RiceBiscuits** [comment]
I think you are almost done. 

1, 3. please see "https://llvm.moe/ocaml/Llvm.html"
2. Clang will link external c function.
4. It depends. Basically same as others language.

**Comment by kws1207** [comment]
There is no search result for llm in "https://llvm.moe/ocaml/Llvm.html". Is llm same with llmodule?

There is also no search result for llctx. Is llctx same with llcontext?

Also in the link, it lacks of detail explanation and says like "See the llvm::Module class.". 
Can you please tell me where can I see a proper explanation?

**Comment by RiceBiscuits** [comment]
Yes. llm and llctx are variable name of llmodule and llcontext. I forgot to explain it. I'm sorry for that.
As you can see, they are containers, and types are abstract type.
I have no idea that explanation of llvm:Module class for Ocaml are existed. 
But if you want to look at more details, there are C++ reference at the LLVM official website. 
(https://llvm.org/doxygen/classllvm_1_1Module.html
https://llvm.org/doxygen/classllvm_1_1LLVMContext.html)

**Comment by kws1207** [comment]
Thank you
